python/tests/expressions/test_expressions_base.py

```diff
@@ -841,3 +924,290 @@ def test_not_expression_binding(unbound_not_expression, expected_bound_expressio
+    """Test visiting an And and In expression with a bound boolean expression visitor"""
+    bound_expression = base.And(
+        base.BoundIn[str](
+            term=base.BoundReference(
+                field=NestedField(field_id=1, name="foo", field_type=StringType(), required=False),
+                accessor=Accessor(position=0, inner=None),
+            ),
+            literals=(StringLiteral("foo"), StringLiteral("bar")),
+        ),
+        base.BoundIn[str](
+            term=base.BoundReference(
+                field=NestedField(field_id=2, name="bar", field_type=StringType(), required=False),
+                accessor=Accessor(position=1, inner=None),
+            ),
+            literals=(StringLiteral("baz"), StringLiteral("qux")),
+        ),
+    )
+    visitor = FooBoundBooleanExpressionVisitor()
+    result = base.visit(bound_expression, visitor=visitor)
+    assert result == ["IN", "IN", "AND"]
+
+
+def test_bound_boolean_expression_visitor_or():
+    """Test visiting an Or expression with a bound boolean expression visitor"""
+    bound_expression = base.Or(
+        base.Not(
+            base.BoundIn[str](
+                base.BoundReference[str](
+                    field=NestedField(field_id=1, name="foo", field_type=StringType(), required=False),
+                    accessor=Accessor(position=0, inner=None),
+                ),
+                {StringLiteral("foo"), StringLiteral("bar")},
+            )
+        ),
+        base.Not(
+            base.BoundIn[str](
+                base.BoundReference[str](
+                    field=NestedField(field_id=2, name="bar", field_type=StringType(), required=False),
+                    accessor=Accessor(position=1, inner=None),
+                ),
+                {StringLiteral("baz"), StringLiteral("qux")},
+            )
+        ),
+    )
+    visitor = FooBoundBooleanExpressionVisitor()
+    result = base.visit(bound_expression, visitor=visitor)
+    assert result == ["IN", "NOT", "IN", "NOT", "OR"]
+
+
+def test_bound_boolean_expression_visitor_equal_and_not_equal():
+    """Test visiting an And expression with a bound boolean expression visitor"""
+    bound_expression = base.And(
+        base.BoundNotEqualTo[str](
+            term=base.BoundReference(
+                field=NestedField(field_id=1, name="foo", field_type=StringType(), required=False),
+                accessor=Accessor(position=0, inner=None),
+            ),
+            literal=StringLiteral("foo"),
+        ),
+        base.BoundEqualTo[str](
+            term=base.BoundReference(
+                field=NestedField(field_id=2, name="bar", field_type=StringType(), required=False),
+                accessor=Accessor(position=1, inner=None),
+            ),
+            literal=StringLiteral("foo"),
+        ),
+    )
+    visitor = FooBoundBooleanExpressionVisitor()
+    result = base.visit(bound_expression, visitor=visitor)
+    assert result == ["NOT_EQUAL", "EQUAL", "AND"]
```

Comment:
Can you add separate tests that validate not-equal and equal in isolation?

Reply:
Done! Much clearer